daft/execution/shuffles/flight_shuffle/flight_shuffle.py

```diff
@@ -0,0 +1,465 @@
+        return file_bytes_per_partition
+
+    def clear_partitions(self, partition_idxs: list[int]):
+        # clean up the shuffle files for the partitions that are being cleared
+        for partition_idx in partition_idxs:
+            path = get_shuffle_file_path(self.node_id, self.shuffle_stage_id, partition_idx)
+            try:
+                if os.path.exists(path):
+                    shutil.rmtree(path)
+            except Exception as e:
+                print(f"failed to clean up shuffle files: {e} for partition {partition_idx}")
+
+    def shutdown(self):
+        # shutdown the server
+        self.server.shutdown()
+        # clean up the shuffle files
+        try:
+            path = get_shuffle_file_path(self.node_id, self.shuffle_stage_id)
+            if os.path.exists(path):
+                shutil.rmtree(path)
+        except Exception as e:
+            print(f"failed to clean up shuffle files: {e}")
+
+
+@ray.remote
+def reduce_partitions(
+    shuffle_stage_id: int,
+    partitions: list[int],
+    shuffle_actor_addresses: list[str],
+    shuffle_actor_manager: ShuffleActorManager,
+    max_parallel_fetches: int = 2,
+):
+    """reduce_partitions is task that produces reduced partitions.
+
+    It is responsible for:
+    - Fetching the partitions from the shuffle actors
+    - Concatenating the partitions
+    - Yielding the concatenated partition and its metadata
+
+    We submit the requests in parallel on a threadpool with an async queue as a backpressure mechanism.
+    """
+    import asyncio
+
+    from daft.dependencies import flight
+
+    # Fetch a single partition from a single actor
+    def fetch_partition_from_actor(client: flight.FlightClient, partition_idx: int):
+        ticket = flight.Ticket(f"{shuffle_stage_id},{partition_idx}".encode())
+        reader = client.do_get(ticket)
+        res = MicroPartition.from_arrow(reader.read_all())
+        return res
+
+    # Fetch all the partitions from all the actors
+    async def fetch_all_partitions(
+        clients: list[flight.FlightClient],
+        partition_idxs: list[int],
+        executor: concurrent.futures.ThreadPoolExecutor,
+        queue: asyncio.Queue,
+    ):
+        for partition_idx in partition_idxs:
+            # Submit all the fetch requests to the thread pool
+            futures = [executor.submit(fetch_partition_from_actor, client, partition_idx) for client in clients]
+            # Put the futures in the queue, the queue is used as a backpressure mechanism
+            await queue.put((partition_idx, futures))
+        # Put None in the queue to indicate that the fetch is done
+        await queue.put(None)
+
+    clients = [flight.FlightClient(address) for address in shuffle_actor_addresses]
+
+    # Generate partitions from the actors
+    async def gen_partitions():
+        loop = asyncio.get_running_loop()
+        with concurrent.futures.ThreadPoolExecutor() as executor:
```

Comment:
we'll end up creating a ton of threads (ncores) for this. We're likely bottleneck by IO anyways so you should be able to use just like 2-4 threads. You can set the max workers in `ThreadPoolExecutor`